core/src/main/java/com/linecorp/armeria/client/RedirectingClient.java

```diff
@@ -0,0 +1,287 @@
+                if (domainFilter != null) {
+                    if (!domainFilter.test(ctx, redirectUri.getHost())) {
+                        endRedirect(ctx, reqDuplicator, responseFuture, response);
+                        return;
+                    }
+                } else if (withBaseUri && !allowSameDomain.test(ctx, redirectUri.getHost())) {
+                    endRedirect(ctx, reqDuplicator, responseFuture, response);
+                    return;
+                }
+            }
+
+            final RequestHeaders newHeaders;
+            try {
+                newHeaders = redirectConfig.redirectRule().shouldRedirect(derivedCtx, requestHeaders,
+                                                                          responseHeaders, redirectUri);
+            } catch (Throwable t) {
+                handleException(ctx, reqDuplicator, responseFuture, t, false);
+                return;
+            }
+
+            if (newHeaders == null) {
+                endRedirect(ctx, reqDuplicator, responseFuture, response);
+                return;
+            }
+
+            handleRedirect(ctx, derivedCtx, redirectCtx, reqDuplicator, response, newHeaders);
+        });
+    }
+
+    private void handleRedirect(ClientRequestContext ctx, ClientRequestContext derivedCtx,
+                                RedirectContext redirectCtx,
+                                HttpRequestDuplicator reqDuplicator, HttpResponse response,
+                                RequestHeaders newHeaders) {
+
+        final CompletableFuture<HttpResponse> responseFuture = redirectCtx.responseFuture();
+        final URI newUri = URI.create(newHeaders.path());
+
+        final String originalPath = redirectCtx.request().path();
+        final String newPath = pathWithQuery(newUri, newUri.getRawQuery());
+
+        if (originalPath.equals(newPath)) {
+            final String newHost = newUri.getHost();
+            final Endpoint endpoint = ctx.endpoint();
+            // endpoint is not null because we already received the response.
+            assert endpoint != null;
+            if (newHost == null || newHost.equals(endpoint.host())) {
+                // Redirect loop!
+                final Set<String> redirectPaths = redirectCtx.redirectPaths();
+                final RedirectLoopsException exception;
+                if (redirectPaths == null) {
+                    exception = new RedirectLoopsException(originalPath);
+                } else {
+                    exception = new RedirectLoopsException(originalPath, redirectPaths);
+                }
+                abortResponse(response, derivedCtx, exception);
+                handleException(ctx, reqDuplicator, responseFuture, exception, false);
+                return;
+            }
+        }
+
+        if (!redirectCtx.addRedirectPath(newUri.toString())) {
+            // Redirect loop!
```

Comment:
Let's say we send a request with POST and a response is 303 with the same path.
Do we need to regard the path as a redirect loop?
```
>>>>>>>>
POST /blogs HTTP/1.1 
...
{
   title: "...",
   content:
}

<<<<<<<
HTTP/1.1 303 SEE OTHER
Location: /blogs
```

Reply:
It shouldn't be considered to be a redirect loop in my opinion.